libretro-common/net/net_http.c

```diff
@@ -761,31 +1385,112 @@ struct http_t *net_http_new(struct http_connection_t *conn)
 #endif
 
       len_str[len] = '\0';
+#ifdef HAVE_DEBUG
+      RARCH_LOG("[HTTP] >> Content-Length: %s\n", len_str);
+#endif
 
       net_http_send_str(&conn->sock_state, &error, len_str);
       net_http_send_str(&conn->sock_state, &error, "\r\n");
 
       free(len_str);
    }
 
+   if (conn->request->headers->len)
+   {
+      for (i = 0;i < conn->request->headers->len;i++)
+      {
+         if (!string_is_equal(conn->request->headers->headers[i].name, "Content-Type") &&
+              !string_is_equal(conn->request->headers->headers[i].name, "User-Agent"))
+         {
+#ifdef HAVE_DEBUG
+            RARCH_LOG("[HTTP] >> %s: %s\n", conn->request->headers->headers[i].name,
+               conn->request->headers->headers[i].value);
+#endif
+            net_http_send_str(&conn->sock_state, &error, conn->request->headers->headers[i].name);
+            net_http_send_str(&conn->sock_state, &error, ": ");
+            net_http_send_str(&conn->sock_state, &error, conn->request->headers->headers[i].value);
+            net_http_send_str(&conn->sock_state, &error, "\r\n");
+         }
+      }
+   }
+
+#ifdef HAVE_DEBUG
+   RARCH_LOG("[HTTP] >> User-Agent: %s\n", useragent ? useragent : "libretro");
+#endif
    net_http_send_str(&conn->sock_state, &error, "User-Agent: ");
-   if (conn->useragentcopy)
-      net_http_send_str(&conn->sock_state, &error, conn->useragentcopy);
-   else
+   if (useragent)
+   {
+      net_http_send_str(&conn->sock_state, &error, useragent);
+   } else
+   {
       net_http_send_str(&conn->sock_state, &error, "libretro");
+   }
    net_http_send_str(&conn->sock_state, &error, "\r\n");
 
+#ifdef HAVE_DEBUG
+   RARCH_LOG("[HTTP] >> Connection: close\n");
+   RARCH_LOG("[HTTP] >>\n");
+#endif
+
    net_http_send_str(&conn->sock_state, &error, "Connection: close\r\n");
    net_http_send_str(&conn->sock_state, &error, "\r\n");
 
-   if (conn->methodcopy && (string_is_equal(conn->methodcopy, "POST")))
-      net_http_send_str(&conn->sock_state, &error, conn->postdatacopy);
+   if (conn->request->method && (string_is_equal(conn->request->method, "POST") ||
+          string_is_equal(conn->request->method, "PUT") || string_is_equal(conn->request->method, "PATCH")))
+   {
+#ifdef HAVE_DEBUG
+      if (conn->request->body.body_type == HTTP_REQUEST_BODY_FROM_RAW)
+      {
+         log_body(true, (char *)conn->request->body.value.raw.data, conn->request->body.value.raw.length);
+      } else
+      {
+         RARCH_LOG("[HTTP] >> ## omitted ##\n");
+      }
+#endif
+
+      switch (conn->request->body.body_type)
+      {
+         case HTTP_REQUEST_BODY_FROM_RAW:
+            net_http_send_bytes(&conn->sock_state, &error, conn->request->body.value.raw.data, conn->request->body.value.raw.length);
+            break;
+         case HTTP_REQUEST_BODY_FROM_FILE:
+         {
+            uint8_t buffer[4096];
+            int64_t bytes_read;
```

Comment:
size_t, same as below

Reply:
`filestream_read()` returns an `int64_t` and it is possible for it to return a negative number.